comp/core/telemetry/telemetry.go

```diff
@@ -0,0 +1,200 @@
+// Unless explicitly stated otherwise all files in this repository are licensed
+// under the Apache License Version 2.0.
+// This product includes software developed at Datadog (https://www.datadoghq.com/).
+// Copyright 2023-present Datadog, Inc.
+
+package telemetry
+
+import (
+	"net/http"
+
+	promOtel "go.opentelemetry.io/otel/exporters/prometheus"
+
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/client_golang/prometheus/collectors"
+	"github.com/prometheus/client_golang/prometheus/promhttp"
+	"go.opentelemetry.io/otel/metric"
+	sdk "go.opentelemetry.io/otel/sdk/metric"
+)
+
+// TODO (components): Remove the global and move this into `newTelemetry` after all telemetry is migrated to the component
+var (
+	registry = newRegistry()
+	provider = newProvider(registry)
+)
+
+type telemetryImpl struct {
+	registry      *prometheus.Registry
+	meterProvider *sdk.MeterProvider
+}
+
+func newRegistry() *prometheus.Registry {
+	reg := prometheus.NewRegistry()
+	reg.MustRegister(collectors.NewProcessCollector(collectors.ProcessCollectorOpts{}))
+	reg.MustRegister(collectors.NewGoCollector())
+	return reg
+}
+
+func newProvider(reg *prometheus.Registry) *sdk.MeterProvider {
+	exporter, err := promOtel.New(promOtel.WithRegisterer(reg))
+
+	if err != nil {
+		panic(err)
+	}
+
+	return sdk.NewMeterProvider(sdk.WithReader(exporter))
+}
+
+func newTelemetry() Component {
+	return &telemetryImpl{
+		registry:      registry,
+		meterProvider: provider,
+	}
+}
+
+// Same as `newTelemetry“ without the global.
+// Can be merged with `newTelemetry` when the global is removed
+func newMock() Component {
+	reg := prometheus.NewRegistry()
+	provider := newProvider(reg)
+
+	return &telemetryImpl{
+		registry:      reg,
+		meterProvider: provider,
+	}
+}
+
+// TODO (components): Remove this when all telemetry is migrated to the component
+func GetCompatComponent() Component {
+	return newTelemetry()
+}
+
+func (t *telemetryImpl) Handler() http.Handler {
+	return promhttp.HandlerFor(t.registry, promhttp.HandlerOpts{})
+}
+
+func (t *telemetryImpl) Reset() {
+	registry = prometheus.NewRegistry()
+	t.registry = registry
+}
+
+func (t *telemetryImpl) Meter(name string, opts ...metric.MeterOption) metric.Meter {
+	return t.meterProvider.Meter(name, opts...)
+}
+
+func (t *telemetryImpl) NewCounter(subsystem, name string, tags []string, help string) Counter {
+	return t.NewCounterWithOpts(subsystem, name, tags, help, DefaultOptions)
+}
+
+func (t *telemetryImpl) NewCounterWithOpts(subsystem, name string, tags []string, help string, opts Options) Counter {
+	name = opts.NameWithSeparator(subsystem, name)
+
+	c := &promCounter{
+		pc: prometheus.NewCounterVec(
+			prometheus.CounterOpts{
+				Subsystem: subsystem,
+				Name:      name,
+				Help:      help,
+			},
+			tags,
+		),
+	}
+	_ = t.registry.Register(c.pc)
```

Comment:
Do you know what kind of error `Register` can  return?

Reply:
[It can return several kinds of errors ](https://github.com/prometheus/client_golang/blob/main/prometheus/registry.go#L294). 

This used to be `MustRegister` which would panic on error. I can change it back if you prefer.

Reply:
It is usually better to not ignore the errors. If `MustRegister` was used so I think it makes sense to keep using it.

Reply:
I made the change in the follow up PR :)